api/server/server.go

```diff
@@ -260,12 +264,74 @@ func NewServer(conf Config) (*Server, error) {
 	}
 
 	s.indexServer = startIndexHTTPServer(s)
+	if err != nil {
+		return nil, fmt.Errorf("starting index server: %s", err)
+	}
+
+	s.ipfsRevProxy, err = startIPFSRevProxy(&conf, ffsManager)
+	if err != nil {
+		return nil, fmt.Errorf("starting IPFS reverse proxy: %s", err)
+	}
 
 	log.Info("Starting finished, serving requests")
 
 	return s, nil
 }
 
+type ffsHTTPAuth struct {
+	cont       http.Handler
+	ffsManager *manager.Manager
+}
+
+func newHTTPFFSAuthInterceptor(cont http.Handler, m *manager.Manager) *ffsHTTPAuth {
+	fha := &ffsHTTPAuth{
+		cont:       cont,
+		ffsManager: m,
+	}
+
+	return fha
+}
+
+func (fha *ffsHTTPAuth) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	authFFS := r.Header.Get("x-ipfs-ffs-auth")
+	_, err := fha.ffsManager.GetByAuthToken(authFFS)
+	if authFFS == "" || err == manager.ErrAuthTokenNotFound {
+		http.Error(rw, "FFS token required", http.StatusUnauthorized)
+		return
+	}
+	fha.cont.ServeHTTP(rw, r)
+}
+
+func startIPFSRevProxy(conf *Config, m *manager.Manager) (*http.Server, error) {
+	if conf.IPFSRevProxyAddr == "" {
+		return nil, nil
+	}
+
+	log.Info("Starting IPFS reverse proxy...")
+	ipfsIP, err := util.TCPAddrFromMultiAddr(conf.IpfsAPIAddr)
+	if err != nil {
+		return nil, fmt.Errorf("converting IPFS multiaddr to tcp addr: %s", err)
+	}
+
+	urlIPFS, err := url.Parse("http://" + ipfsIP)
+	if err != nil {
+		return nil, fmt.Errorf("generating IPFS URL for reverse proxy: %s", err)
+	}
+	rph := httputil.NewSingleHostReverseProxy(urlIPFS)
+	fha := newHTTPFFSAuthInterceptor(rph, m)
+	rp := &http.Server{
+		Addr:    conf.IPFSRevProxyAddr,
+		Handler: fha,
+	}
+
+	go func() {
+		if err := rp.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			log.Fatalf("serving index http: %v", err)
+		}
+	}()
+	return rp, nil
+}
+
```

Comment:
So here is the reverse proxy in `powd`, which is only enabled if a non-empty `IPFSRevProxyAddr` was provided on startup.

If that's the case, it creates a reverse proxy to the IPFS HTTP API with an extra tweak of enforcing a valid FFS token is provided in all HTTP calls.